src/fairchem/core/common/relaxation/optimizers/optimizable.py

```diff
@@ -0,0 +1,560 @@
+
+    def get_cells(self) -> torch.Tensor:
+        """Get batch crystallographic cells."""
+        return self.batch.cell
+
+    def set_cells(self, cells: torch.Tensor | NDArray) -> None:
+        """Set batch cells."""
+        assert self.batch.cell.shape == cells.shape, "Cell shape mismatch"
+        if isinstance(cells, np.ndarray):
+            cells = torch.tensor(cells, dtype=torch.float32, device=self.device)
+        cells = cells.to(dtype=torch.float32, device=self.device)
+        self.batch.cell[self.update_mask] = cells[self.update_mask]
+
+    def get_volumes(self) -> torch.Tensor:
+        """Get a tensor of volumes for each cell in batch"""
+        cells = self.get_cells()
+        return torch.linalg.det(cells)
+
+    def iterimages(self) -> Batch:
+        # XXX document purpose of iterimages - this is just needed to work with ASE optimizers
+        yield self.batch
+
+    def get_max_forces(
+        self, forces: torch.Tensor | None = None, apply_constraint: bool = False
+    ) -> torch.Tensor:
+        """Get the maximum forces per structure in batch"""
+        if forces is None:
+            forces = self.get_forces(apply_constraint=apply_constraint, no_numpy=True)
+        return scatter((forces**2).sum(axis=1).sqrt(), self.batch_indices, reduce="max")
+
+    def converged(
+        self,
+        forces: torch.Tensor | NDArray | None,
+        fmax: float,
+        max_forces: torch.Tensor | None = None,
+    ) -> bool:
+        """Check if norm of all predicted forces are below fmax"""
+        if forces is not None:
+            if isinstance(forces, np.ndarray):
+                forces = torch.tensor(forces, device=self.device, dtype=torch.float32)
+            max_forces = self.get_max_forces(forces)
+        elif max_forces is None:
+            max_forces = self.get_max_forces()
+
+        update_mask = max_forces.ge(fmax)
+        # update cached mask
+        if self.mask_converged:
+            if self._update_mask is None or self._cumulative_mask is False:
+                self._update_mask = update_mask
+            else:
+                # some models can have random noise in their predictions, so the mask is updated by
+                # keeping all previously converged structures masked even if new force predictions
+                # push it slightly above threshold
+                self._update_mask = torch.logical_and(self._update_mask, update_mask)
+            update_mask = self._update_mask
+
+        return not torch.any(update_mask).item()
+
+    def get_atoms_list(self) -> list[Atoms]:
+        """Get ase Atoms objects corresponding to the batch"""
+        self._predict()  # in case no predictions have been run
+        return batch_to_atoms(self.batch, results=self.torch_results)
+
+    def update_graph(self):
+        """Update the graph if model does not use otf_graph."""
+        edge_index, cell_offsets, num_neighbors = radius_graph_pbc(self.batch, 6, 50)
```

Comment:
Does it make sense to use `self.trainer.model.generate_graph`instead? 

https://github.com/FAIR-Chem/fairchem/blob/f19c51b66b43edeff5649641a2f57b1581a8c6b0/src/fairchem/core/models/base.py#L4